From 70352471132c29bf7e5e97e0c38c5e97ab2cf14a Mon Sep 17 00:00:00 2001
From: CI Assistant <chdauto@intel.com>
Date: Tue, 7 Apr 2020 13:56:14 +0300
Subject: [PATCH] Merge pull request #2022 in SW_UGW/linux from
 bugfix/UGW_SW-49016-qos-wmm-marking-fix-8.4.1 to 8.4.1

* commit '2686ce755f44fcad50ead416cd6948b73a0ac7ec':
  UGW_SW-49016 : PPA/QoS class2prio mapping notify fix
---
 drivers/net/ethernet/lantiq/ppa/ppa_hook.c | 35 +++++++++++++++++++++++++-----
 include/net/ppa/ppa_api.h                  |  4 ----
 include/net/ppa/ppa_hook.h                 | 20 ++++++++++++-----
 include/uapi/net/ppa_api.h                 |  5 +++++
 4 files changed, 49 insertions(+), 15 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/ppa/ppa_hook.c b/drivers/net/ethernet/lantiq/ppa/ppa_hook.c
index bc4fe8bbaf1a..0b05c623ba86 100644
--- a/drivers/net/ethernet/lantiq/ppa/ppa_hook.c
+++ b/drivers/net/ethernet/lantiq/ppa/ppa_hook.c
@@ -253,9 +253,35 @@ int32_t (*ppa_check_if_netif_fastpath_fn)(PPA_NETIF *, char *, uint32_t) = NULL;
 int32_t (*ppa_hook_disconn_if_fn)(PPA_NETIF *, PPA_DP_SUBIF *, uint8_t *, uint32_t) = NULL;
 
 #if defined(WMM_QOS_CONFIG) && WMM_QOS_CONFIG
-int32_t (*ppa_register_qos_class2prio_hook_fn)(int32_t, PPA_NETIF *, PPA_QOS_CLASS2PRIO_CB, uint32_t) = NULL;
-#endif
-#endif
+static RAW_NOTIFIER_HEAD(ppa_event_chain);
+
+int ppa_register_event_notifier(struct notifier_block *nb)
+{
+	return raw_notifier_chain_register(&ppa_event_chain, nb);
+}
+EXPORT_SYMBOL(ppa_register_event_notifier);
+
+int ppa_unregister_event_notifier(struct notifier_block *nb)
+{
+	return raw_notifier_chain_unregister(&ppa_event_chain, nb);
+}
+EXPORT_SYMBOL(ppa_unregister_event_notifier);
+
+int ppa_call_class2prio_notifiers(unsigned long val,
+				  s32 port_id, PPA_NETIF *dev,
+				  u8 class2prio[MAX_TC_NUM])
+{
+	struct ppa_class2prio_notifier_info info;
+
+	info.port_id = port_id;
+	info.dev = dev;
+	memcpy(info.class2prio, class2prio, (sizeof(u8) * MAX_TC_NUM));
+
+	return raw_notifier_call_chain(&ppa_event_chain, val, &info);
+}
+EXPORT_SYMBOL(ppa_call_class2prio_notifiers);
+#endif /* WMM_QOS_CONFIG */
+#endif /* CONFIG_PPA_API_DIRECTCONNECT */
 
 #if IS_ENABLED(CONFIG_PPA_API_DIRECTPATH)
 int32_t (*ppa_directpath_port_add_fn)(void) = NULL;
@@ -486,9 +512,6 @@ EXPORT_SYMBOL(ppa_hook_get_netif_accel_stats_fn);
 #if IS_ENABLED(CONFIG_PPA_API_DIRECTCONNECT)
 EXPORT_SYMBOL(ppa_check_if_netif_fastpath_fn);
 EXPORT_SYMBOL(ppa_hook_disconn_if_fn);
-#if defined(WMM_QOS_CONFIG) && WMM_QOS_CONFIG
-EXPORT_SYMBOL(ppa_register_qos_class2prio_hook_fn);
-#endif
 #endif
 
 #if IS_ENABLED(CONFIG_PPA_API_DIRECTPATH)
diff --git a/include/net/ppa/ppa_api.h b/include/net/ppa/ppa_api.h
index ebc0ac299ddd..43b59561e93e 100644
--- a/include/net/ppa/ppa_api.h
+++ b/include/net/ppa/ppa_api.h
@@ -75,10 +75,6 @@
  * Definition
  * ####################################
  */
-#if defined(WMM_QOS_CONFIG) && WMM_QOS_CONFIG
-#define WMM_QOS_DEV_F_REG			0x00000001
-#define WMM_QOS_DEV_F_DREG			0x00000002
-#endif/* WMM_QOS_CONFIG */
 #ifdef NO_DOXY
 #define VLAN_ID_SPLIT(full_id, pri, cfi, vid)	do { pri = ((full_id) >> 13) & 7; cfi = ((full_id) >> 12) & 1; vid = (full_id) & 0xFFF } while (0)
 #define VLAN_ID_CONBINE(full_id, pri, cfi, vid) full_id	= (((uint16_t)(pri) & 7) << 13) | (((uint16_t)(cfi) & 1) << 12) | ((uint16_t) (vid) & 0xFFF)
diff --git a/include/net/ppa/ppa_hook.h b/include/net/ppa/ppa_hook.h
index b28b6ff99f66..ff390a737faf 100644
--- a/include/net/ppa/ppa_hook.h
+++ b/include/net/ppa/ppa_hook.h
@@ -233,14 +233,24 @@ extern int32_t ppa_check_if_netif_fastpath_fn(PPA_NETIF *netif, char *ifname, ui
  */
 #ifdef NO_DOXY
 extern int32_t (*ppa_hook_disconn_if_fn)(PPA_NETIF *dev, PPA_DP_SUBIF *subif, uint8_t *mac, uint32_t flags);
-#if defined(WMM_QOS_CONFIG) && WMM_QOS_CONFIG
-extern int32_t (*ppa_register_qos_class2prio_hook_fn)(int32_t port_id, PPA_NETIF *netif, PPA_QOS_CLASS2PRIO_CB qos_class2prio_cb, uint32_t flags);
-#endif /* WMM*/
 #else
 extern int32_t ppa_hook_disconn_if_fn(PPA_NETIF *dev, PPA_DP_SUBIF *subif, uint8_t *mac, uint32_t flags);
+#endif
 #if defined(WMM_QOS_CONFIG) && WMM_QOS_CONFIG
-extern int32_t ppa_register_qos_class2prio_hook_fn(int32_t port_id, PPA_NETIF *netif, PPA_QOS_CLASS2PRIO_CB qos_class2prio_cb, uint32_t flags);
-#endif /* WMM*/
+/* PPA class2prio notifier chain. */
+#define PPA_CLASS2PRIO_DEFAULT	0x0001 /* Notify class2prio default */
+#define PPA_CLASS2PRIO_CHANGE	0x0002 /* Notify class2prio change */
+
+struct ppa_class2prio_notifier_info {
+	s32 port_id;
+	PPA_NETIF *dev;
+	u8 class2prio[MAX_TC_NUM];
+};
+extern int ppa_register_event_notifier(PPA_NOTIFIER_BLOCK *nb);
+extern int ppa_unregister_event_notifier(PPA_NOTIFIER_BLOCK *nb);
+extern int ppa_call_class2prio_notifiers(unsigned long val,
+					 s32 port_id, struct net_device *dev,
+					 u8 class2prio[MAX_TC_NUM]);
 #endif
 #endif
 /*!
diff --git a/include/uapi/net/ppa_api.h b/include/uapi/net/ppa_api.h
index d9bcd943d539..9776c62c1a4c 100755
--- a/include/uapi/net/ppa_api.h
+++ b/include/uapi/net/ppa_api.h
@@ -365,6 +365,11 @@
 	\brief MAX_PRIO_NUM */
 #define MAX_PRIO_NUM 8
 
+/*!
+ *	\brief MAX_WLAN_DEV
+ */
+#define MAX_WLAN_DEV 4
+
 #define TC_START_BIT_POS		0
 #define TC_MASK				 0x1f
 /*!
